bounded-collections/src/const_int.rs

```diff
@@ -0,0 +1,160 @@
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+use crate::{Get, TypedGet};
+
+// Numbers which have constant upper and lower bounds.
+trait ConstBounded<T> {
+	const MIN: T;
+	const MAX: T;
+}
+
+macro_rules! impl_const_bounded {
+	($bound:ty, $t:ty) => {
+		impl ConstBounded<$bound> for $t {
+			const MIN: $bound = <$t>::MIN as $bound;
+			const MAX: $bound = <$t>::MAX as $bound;
+		}
+	};
+}
+
+impl_const_bounded!(u128, u8);
+impl_const_bounded!(u128, u16);
+impl_const_bounded!(u128, u32);
+impl_const_bounded!(u128, u64);
+impl_const_bounded!(u128, u128);
+impl_const_bounded!(u128, usize);
+
+impl_const_bounded!(i128, i8);
+impl_const_bounded!(i128, i16);
+impl_const_bounded!(i128, i32);
+impl_const_bounded!(i128, i64);
+impl_const_bounded!(i128, i128);
+
+// Check whether a number is within the bounds of a type.
+struct CheckOverflow<T>(core::marker::PhantomData<T>);
+
+impl CheckOverflow<u128> {
+	const fn check<T: ConstBounded<u128>, const N: u128>() {
+		assert!(N <= T::MAX && N >= T::MIN)
+	}
+}
+
+impl CheckOverflow<i128> {
+	const fn check<T: ConstBounded<i128>, const N: i128>() {
+		assert!(N <= T::MAX && N >= T::MIN)
+	}
+}
+
+/// Const getter for unsigned integers.
+#[derive(Default, Clone)]
+pub struct ConstUint<const N: u128>;
+
+#[cfg(feature = "std")]
+impl<const N: u128> std::fmt::Debug for ConstUint<N> {
+	fn fmt(&self, fmt: &mut std::fmt::Formatter) -> std::fmt::Result {
+		fmt.write_str(&format!("{}<{}>", stringify!(ConstUint), N))
+	}
+}
+#[cfg(not(feature = "std"))]
+impl<const N: u128> core::fmt::Debug for ConstUint<N> {
+	fn fmt(&self, fmt: &mut core::fmt::Formatter) -> core::fmt::Result {
+		fmt.write_str("<wasm:stripped>")
+	}
+}
+
+impl<const N: u128> TypedGet for ConstUint<N> {
+	type Type = u128;
+	fn get() -> u128 {
+		N
+	}
+}
+
+/// Const getter for signed integers.
+#[derive(Default, Clone)]
+pub struct ConstInt<const N: i128>;
+
+#[cfg(feature = "std")]
+impl<const N: i128> std::fmt::Debug for ConstInt<N> {
+	fn fmt(&self, fmt: &mut std::fmt::Formatter) -> std::fmt::Result {
+		fmt.write_str(&format!("{}<{}>", stringify!(ConstInt), N))
+	}
+}
+#[cfg(not(feature = "std"))]
+impl<const N: i128> core::fmt::Debug for ConstInt<N> {
+	fn fmt(&self, fmt: &mut core::fmt::Formatter) -> core::fmt::Result {
+		fmt.write_str("<wasm:stripped>")
+	}
+}
+
+impl<const N: i128> TypedGet for ConstInt<N> {
+	type Type = i128;
+	fn get() -> i128 {
+		N
+	}
+}
+
+macro_rules! impl_const_int {
+	($t:ident, $bound:ty, $target:ty) => {
+		impl<const N: $bound> Get<$target> for $t<N> {
+			fn get() -> $target {
+				let _ = <CheckOverflow<$bound>>::check::<$target, N>();
+				N as $target
+			}
+		}
+		impl<const N: $bound> Get<Option<$target>> for $t<N> {
+			fn get() -> Option<$target> {
+				let _ = <CheckOverflow<$bound>>::check::<$target, N>();
```

Comment:
This is just a runtime check no?

If I write:
```rust
pub fn foo() -> u8 {
	<ConstUint<4200> as Get<u8>>::get()
}
```
it will compile fine no?

Reply:
You're absolutely right. Initially, I wrote code that worked correctly, but in trying to refactor to make it more concise, overlooked this detail. Thanks for pointing out!